File: Study04/Dijkstra.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Scanner;

public class Dijkstra {

	public static Node[] nodes = null;
	public static PriorityQueue p_queue = null;
	public static int start;
	public static int goal;

	public static void main(String[] args) {
		// 入力の読み込み
		start = Integer.parseInt(args[0]);
		goal = Integer.parseInt(args[1]);
		nodes = getNodeList();
		p_queue = new PriorityQueue();

		p_queue.enqueue(nodes[start]);
		// 各ノードに注目しながら隣接ノードのコストを更新していく
		while(!p_queue.isEmpty()) {
			Node focused_node = p_queue.dequeue();
			if(focused_node.isGoal()) break;
			updateCost(focused_node);
		}
		System.out.println(getRoute(nodes[goal]));
	}

	private static void updateCost(Node node) {
		// 注目済みでないことをチェック
		if(!node.is_used) {
			// 各エッジについて順次に調べる
			for(int key : node.edge_cost.keySet()) {
				// エッジの接続先ノードが注目済みであるか
				if(!nodes[key].is_used) {
					int tmp = node.node_cost + node.edge_cost.get(key);
					if(nodes[key].node_cost > tmp) {
						nodes[key].node_cost = tmp;
						nodes[key].prev_node = node;
					}
					p_queue.enqueue(nodes[key]);
				}
			}
			node.is_used = true;
		}
		System.out.println(p_queue.toString());
	}

	// 出発ノードから目的ノードまでのルートを文字列として得る
	private static String getRoute(Node node) {
		if(node.isStart()) return node.id + "";
		else {
			if(node.prev_node != null) {
				if(node.isGoal()) return "最小コスト : " + node.node_cost + "\n" + getRoute(node.prev_node) + " - "+ node.id;
				else return getRoute(node.prev_node) + " - " + node.id;
			} else { 
				return "目的地に到達できません";
			}
		}
	}

	// ファイルからデータを読み取る
	private static Node[] getNodeList() {
		Scanner scanner = new Scanner(System.in);
		LinkedList<Node> tmp_nodes = new LinkedList<Node>();
		int i = 0, count_edge = 0;
		while(scanner.hasNextLine()) {
			String[] edges = scanner.nextLine().split(",");
			HashMap<Integer, Integer> tmp_hash = new HashMap<Integer, Integer>();
			for(String edge : edges) {
				String[] pair = edge.split(":");
				tmp_hash.put(Integer.parseInt(pair[0]), Integer.parseInt(pair[1]));
				count_edge++;
			}
			tmp_nodes.add(new Node(i, tmp_hash, 2));
			i++;
		}
    System.out.println("ノード数 : " + tmp_nodes.size());
    System.out.println("エッジ数 : " + count_edge);
    tmp_nodes.get(start).kind = 0;
    tmp_nodes.get(start).node_cost = 0;
    tmp_nodes.get(goal).kind = 1;
		return tmp_nodes.toArray(new Node[tmp_nodes.size()]);
	}

}

// ノードを表現するクラス
class Node {

	int id, node_cost;
	// <エッジの接続先, エッジのコスト>
	HashMap<Integer, Integer> edge_cost;
	// 0:出発ノード, 1:ゴールノード, 2:その他, -1:空(優先度付きキューで使用)
	int kind;
	boolean is_used;
	Node prev_node;

	public Node(int id, HashMap<Integer, Integer> edge_cost, int kind) {
		this.id = id;
		this.edge_cost = edge_cost;
		this.node_cost = Integer.MAX_VALUE;
		this.kind = kind;
		this.is_used = false;
		this.prev_node = null;
	}
	public boolean isStart() { return this.kind == 0; }
	public boolean isGoal() { return this.kind == 1; }
	public boolean isNone() { return this.kind == -1; }

}

// 優先度付きキュー
class PriorityQueue {

	// リスト構造でヒープを再現
	private LinkedList<Node> heap = new LinkedList<Node>();

	public static final Node NONE = new Node(-1, null, -1);

	public void enqueue(Node node) {
		if(!heap.contains(NONE)) {
			heap.add(node);
			upheap(node, heap.size()-1);
		} else {
			int index = heap.indexOf(NONE);
			upheap(node, index);
		}
	}
	private void upheap(Node node, int i) {
		if(i == 0) heap.set(i, node);
		else {
			Node parent = heap.get((i+1)/2-1);
			if(node.node_cost >= parent.node_cost) heap.set(i, node);
			else {
				heap.set(i, parent);
				upheap(node, (i+1)/2-1);
			}
		}
	}

	public Node dequeue() {
		Node head = heap.get(0);
		heap.set(0, NONE);
		downheap(0);
		return head;
	}
	private void downheap(int i) {
		int j  = (i+1)*2-1;
		if(j < heap.size()) {
			if(j+1 < heap.size()) {
				if(heap.get(j).node_cost > heap.get(j+1).node_cost)  j++;
			}
			if(heap.get(i).node_cost > heap.get(j).node_cost) {
				Node tmp = heap.get(i);
				heap.set(i, heap.get(j));
				heap.set(j, tmp);
				downheap(j);
			}
		}
	}

	public boolean isEmpty() { 
		for(Node node : heap) { if(!node.isNone()) return false; }
		return true;
	}
	@Override
	public String toString() {
		String str = "";
		for(Node node : heap) {
			str += !node.isNone() ? (node.id+":"+node.node_cost+", ") : "none, ";
		}
		return str;
	}
}

File: Study04/Dijkstra.scala
// scala source code

import scala.io.Source

object Dijkstra {

  var nodes :Seq[Node] = null
  var p_queue :PriorityQueue = new PriorityQueue()

  def main(args :Array[String]) {
    // 各値の取得
    val start = args(0).toInt
    val goal = args(1).toInt
    nodes = Node.getNodeSeq(Source.stdin.getLines, start, goal)

    p_queue.enqueue(nodes(start))
    updateCost(p_queue.dequeue)
    println(getRoute(nodes(goal)))
  }

  // 次のノードを調べてコスト更新
  def updateCost(node :Node) :Unit = {
    if(!node.is_used) {
      node.edge_cost.keys.filter{e => !nodes(e).is_used}.foreach { key =>
        // 注目ノードから接続できるで更新できるノードであれば更新する
        val tmp = node.node_cost + node.edge_cost(key)
        if(nodes(key).node_cost > tmp) {
          nodes(key).node_cost = tmp
          nodes(key).prev_node = Some(node)
        }
        p_queue.enqueue(nodes(key))
      }
      node.is_used = true
    }
    println(p_queue.toString)
    // (注目ノードが目的地でない && 優先度付きキューが空でない) ならば再起を続ける
    if(!node.isGoal && !p_queue.isEmpty) updateCost(p_queue.dequeue)
  }
  
  // 目的地から出発地まで戻りながらルートを見ていく
  def getRoute(node :Node) :String = {
    if(node.isStart) s"${node.id}"
    else {
      node.prev_node match {
        case Some(n) => {
          if(node.isGoal) s"最小コスト : ${node.node_cost}\n${getRoute(n)} - ${node.id}"
          else s"${getRoute(n)} - ${node.id}"
        }
        case None => s"目的地に到達できません - ${node.id}"
      }
    }
  }

}

// edge_cost = [next node, edge's cost]
// kind = 0:start, 1:goal, 2:others, -1:empty
case class Node(val id :Int,val edge_cost :Map[Int, Int], var kind :Short = 2, var node_cost :Int = Int.MaxValue, var is_used :Boolean = false, var prev_node :Option[Node] = None) {

  def isGoal() :Boolean = {
    kind == 1
  }

  def isStart() :Boolean = {
    kind == 0
  }

}

object Node {

  // 入力のパース
  def getNodeSeq(input :Iterator[String], start :Int, goal :Int) :Seq[Node] = {
    var i = -1
    var count_edge = 0
    val nodes = input.map{ line => 
      i+=1
      val node_info = line.split(',')
      val edges = Map.newBuilder[Int, Int]
      (0 until node_info.length).foreach{ i => 
        count_edge += 1
        val edge = node_info(i).split(':')
        edges += edge(0).toInt -> edge(1).toInt
      }
      Node(i, edges.result)
    }.toSeq
    println(s"ノード数 : ${nodes.size}")
    println(s"エッジ数 : ${count_edge}")
    nodes(start).kind = 0
    nodes(start).node_cost = 0
    nodes(goal).kind = 1
    nodes
  }

}

// 根ノードがノードコスト最小となるヒープで再現
class PriorityQueue() {

  // リストでヒープを表現
  var heap :Seq[Node] = Seq.empty
  // 空ノードを表現
  val none = Node(-1, null, -1)

  def enqueue(node :Node) :Unit = {
    if(!heap.contains(none)) {
      // 空ノードがなければリストに新しい枠を追加
      heap = upheap(node, heap :+ node, heap.length)
    } else {
      // 空ノードがあればにenqueueしたいノードを代入
      val i = heap.indexOf(none)
      heap = upheap(node, heap.updated(i, node), i)
    }
  }
  def upheap(node :Node, seq :Seq[Node], i :Int) :Seq[Node] = {
    if(i == 0) {
      // 根ノードまで到達したので代入して再起を止める
      seq.updated(i, node)
    }else {
      val parent = seq((i+1)/2-1)
      // 子ノードの方がコストが大きいので再起を止める
      if(node.node_cost >= parent.node_cost) seq.updated(i, node)
      // 子ノードの方がコストが小さいので再起を続ける
      else upheap(node, seq.updated(i, parent), (i+1)/2-1)
    }
  }

  def dequeue() :Node = {
    // 先頭要素(根ノード, 最小コストのノード)を取り出す
    val node :Node = heap.head
    heap = downheap(heap.updated(0, none))
    node
  }
  def downheap(seq :Seq[Node], i :Int = 0) :Seq[Node] = {
    // j は子ノードのインデックス
    var j = (i+1)*2-1
    if(j < seq.length) {
      if(j+1 < seq.length) {
        if(seq(j).node_cost > seq(j+1).node_cost) j += 1
      }
      // 子ノードとの入れ替えが生じるか
      if(seq(i).node_cost <= seq(j).node_cost) seq
      else downheap(seq.updated(i, seq(j)).updated(j, seq(i)), j)
    } else {
      // 子ノードがない
      seq
    }
  }

  def isEmpty() :Boolean = { heap.filter{e => e != none}.isEmpty }

  override def toString() :String = {
    s"${heap.map{e => s"${if(e != none) s"${e.id}:${e.node_cost}" else "none"}"}.toString()}"
  }
}
